moveit_py/src/moveit_py/pybind11_utils/include/moveit_py/pybind11_utils/ros_msg_typecasters.h

```diff
@@ -0,0 +1,124 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2022, Peter David Fagan, Robert Haschke
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * The name of Robert Haschke may not be use to endorse or promote
+ *     products derived from this software without specific prior
+ *     written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Peter David Fagan, Robert Haschke */
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/serialization.hpp>
+#include <rclcpp/duration.hpp>
+
+namespace py = pybind11;
+
+namespace pybind11
+{
+namespace detail
+{
+
+// Base class for type conversion (C++ <-> python) of ROS message types
+template <typename T>
+struct RosMsgTypeCaster
+{
+  // C++ -> Python
+  // TODO: add error handling
+  static handle cast(const T& src, return_value_policy /* policy */, handle /* parent */)
+  {
+    // serialize src
+    rclcpp::Serialization<T> serializer;
+    rclcpp::SerializedMessage serialized_msg;
+    serializer.serialize_message(&src, &serialized_msg);
+    py::bytes bytes = py::bytes(reinterpret_cast<const char*>(serialized_msg.get_rcl_serialized_message().buffer),
+                                serialized_msg.get_rcl_serialized_message().buffer_length);
+
+    // get Python object type
+    const std::string ros_msg_name = rosidl_generator_traits::data_type<T>();
```

Comment:
Use `rosidl_generator_traits::name<T>();` instead and you can search for single slash (`/`) instead of double colon:
```cpp
// split <package>.msg.<class> into <package>.msg and <class>
char* split_module_and_type(char* module_name)
{
  // replace first slash with dot
  char* sep = strchr(module_name, '/');
  *sep++ = '.';
  // replace second slash with \0
  sep = strchr(sep, '/');
  *sep++ = '\0';
  return sep;
}

py::handle pyclass_from_name(const char* msg_name)
{
  char* module_name = strdup(msg_name);
  char* class_name = split_module_and_type(module_name);

  auto module = py::module::import(module_name);
  auto pyclass = module.attr(class_name);

  free(module_name);
  return pyclass;
}
```

Reply:
eaef7b0